src/store/reducers/macro.ts

```diff
@@ -1,16 +1,149 @@
+    let newMacro: Macro;
+
     switch (action.type) {
-        case MacroActions.GET_ALL:
-            break;
+        case MacroActions.DUPLICATE:
+
+            newMacro = new Macro(action.payload);
+            newMacro.name = generateName(state, newMacro.name);
+            newMacro.id = generateId(state);
+
+            return [...state, newMacro];
+
+        case MacroActions.EDIT_NAME:
+            let name: string = generateName(state, action.payload.name);
+
+            return state.map((macro: Macro) => {
+                if (macro.id === action.payload.id) {
+                    macro.name = name;
+                }
+
+                return macro;
+            });
+
+        case MacroActions.REMOVE:
+            return state.filter((macro: Macro) => macro.id !== action.payload);
+
+        case MacroActions.ADD_ACTION:
+            return state.map((macro: Macro) => {
+                if (macro.id === action.payload.id) {
+                    newMacro = new Macro(macro);
+                    newMacro.macroActions.elements.push(action.payload.action);
+
+                    return newMacro;
+                }
+
+                return macro;
+            });
+
+        case MacroActions.SAVE_ACTION:
+            return state.map((macro: Macro) => {
+                if (macro.id === action.payload.id) {
+                    newMacro = new Macro(macro);
+                    newMacro.macroActions.elements[action.payload.index] = action.payload.action;
+
+                    return newMacro;
+                }
+
+                return macro;
+            });
+
+        case MacroActions.DELETE_ACTION:
+            return state.map((macro: Macro) => {
+                if (macro.id === action.payload.id) {
+                    newMacro = new Macro(macro);
+                    newMacro.macroActions.elements.splice(action.payload.index, 1);
+
+                    return newMacro;
+                }
+
+                return macro;
+            });
+
+        case MacroActions.REORDER_ACTION:
+            return state.map((macro: Macro) => {
+                if (macro.id === action.payload.id) {
+                    let newIndex: number = action.payload.newIndex;
+
+                    // We need to reduce the new index for one when we are moving action down
+                    if (newIndex > action.payload.oldIndex) {
+                        --newIndex;
+                    }
+
+                    newMacro = new Macro(macro);
+                    newMacro.macroActions.elements.splice(
+                        newIndex,
+                        0,
+                        newMacro.macroActions.elements.splice(action.payload.oldIndex, 1)[0]
+                    );
+
+                    return newMacro;
+                }
+
+                return macro;
+            });
+
         default: {
             return state;
         }
     }
 }
+
+export function getMacro(id: number) {
+    // Convert to real number
+    id = +id;
```

Comment:
What do you mean real number? If the parameter is string, then the caller should convert it to number.